File: run_conformance.py
#!/usr/bin/env python3
"""
run_conformance.py — the conformance bar.

Asserts every fixture verifies to its DECLARED expectation:
  - the positive fixture passes all three joined suites end to end;
  - each negative fixture fails for EXACTLY ONE reason, and that reason matches its declared
    expected_failure_reason (and every other suite still passes — one broken join, not several).

Exit 0 iff the whole suite meets the bar. Run from anywhere:
    python3 run_conformance.py
"""
from __future__ import annotations

import json
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE))
from verifier import verify_fixture  # noqa: E402

FIX = HERE / "fixtures"


def main() -> int:
    fixtures = sorted(p for p in FIX.glob("*.json")
                      if p.name not in ("trust_policy.json", "live_confirmed_anchor.json"))
    failures = []
    for path in fixtures:
        fx = json.loads(path.read_text())
        r = verify_fixture(fx)
        exp_pass = fx["expected_overall"] == "pass"
        exp_reason = fx.get("expected_failure_reason")

        # 1) overall verdict matches
        if r["overall_pass"] != exp_pass:
            failures.append(f"{path.name}: overall_pass={r['overall_pass']} expected {exp_pass}")
            continue
        if exp_pass:
            print(f"  ✓ {path.name}: PASS (all three suites join)")
            continue

        # 2) negative: fails for exactly the declared reason
        if r["failure_reason"] != exp_reason:
            failures.append(f"{path.name}: failed with {r['failure_reason']!r}, expected {exp_reason!r}")
            continue
        # 3) exactly ONE broken join (every other suite passes)
        broken = [name for name, s in r["suites"].items() if not s["pass"]]
        if len(broken) != 1:
            failures.append(f"{path.name}: {len(broken)} broken suites {broken}, expected exactly 1")
            continue
        print(f"  ✓ {path.name}: FAIL ({exp_reason}) — exactly one broken join, others pass")

    print("-" * 64)
    if failures:
        print(f"CONFORMANCE BAR NOT MET — {len(failures)} issue(s):")
        for f in failures:
            print(f"    ✗ {f}")
        return 1
    print(f"CONFORMANCE BAR MET — {len(fixtures)} fixtures verify to their declared expectations.")
    print("Positive proves the three-suite join end to end; each negative fails for exactly one reason.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/independent_checker_js/run_conformance.js
#!/usr/bin/env node
/**
 * run_conformance.js — the conformance bar, asserted with THIS independent JS checker.
 *
 * Same assertion shape as run_conformance.py (the original Python runner), but every recompute
 * below goes through verify.js, not verifier.py — this is the CI gate that keeps the second
 * checker honest: if this file and run_conformance.py ever disagree on a fixture's declared
 * expected_overall/expected_failure_reason, CI goes red on THIS runner without needing the
 * Python one to also fail, proving the two are genuinely checked independently, not one silently
 * riding on the other's green.
 *
 * Exit 0 iff every fixture matches its own declared expectation exactly.
 */
'use strict';

const fs = require('fs');
const path = require('path');
const { verifyFixture } = require('./verify.js');

const FIXTURES_DIR = path.join(__dirname, '..', '..', 'fixtures');
const SKIP = new Set(['trust_policy.json', 'live_confirmed_anchor.json']);

function main() {
  const files = fs.readdirSync(FIXTURES_DIR)
    .filter((f) => f.endsWith('.json') && !SKIP.has(f))
    .sort();

  const failures = [];
  for (const file of files) {
    const fx = JSON.parse(fs.readFileSync(path.join(FIXTURES_DIR, file), 'utf8'));
    const r = verifyFixture(fx);
    const expectPass = fx.expected_overall === 'pass';
    const expectReason = fx.expected_failure_reason || null;

    if (r.overall_pass !== expectPass) {
      failures.push(`${file}: overall_pass=${r.overall_pass} expected ${expectPass}`);
      continue;
    }
    if (expectPass) {
      console.log(`  ✓ ${file}: PASS (all five invariants join)`);
      continue;
    }
    if (r.failure_reason !== expectReason) {
      failures.push(`${file}: failed with ${JSON.stringify(r.failure_reason)}, expected ${JSON.stringify(expectReason)}`);
      continue;
    }
    const broken = Object.entries(r.suites).filter(([, s]) => s.pass === false).map(([n]) => n);
    if (broken.length !== 1) {
      failures.push(`${file}: ${broken.length} broken suites [${broken.join(', ')}], expected exactly 1`);
      continue;
    }
    console.log(`  ✓ ${file}: FAIL (${expectReason}) — exactly one broken join, others pass`);
  }

  console.log('-'.repeat(64));
  if (failures.length) {
    console.log(`CONFORMANCE BAR NOT MET (independent JS checker) — ${failures.length} issue(s):`);
    for (const f of failures) console.log(`    ✗ ${f}`);
    process.exit(1);
  }
  console.log(`CONFORMANCE BAR MET (independent JS checker) — ${files.length} fixtures verify to their declared expectations.`);
  console.log('Recomputed from spec text + fixture bytes alone, in a different language, importing');
  console.log('nothing from verifier.py or _bip340_nostr.py — a real second checker, not an adapter.');
}

main();
